sharkfuser/include/fusilli/extras.h

```diff
@@ -0,0 +1,244 @@
+                            "Failed to read file: " + path.string());
+
+    return ok(buffer);
+  }
+
+  // Utility method to build the path to cache file given `graphName` and
+  // `fileName`.
+  //
+  // Format: $HOME/.cache/fusilli/<sanitized version of graphName>/<fileName>
+  static std::filesystem::path getPath(const std::string &graphName,
+                                       const std::string &fileName);
+
+private:
+  // Class should be constructed using one of the factory functions.
+  CacheFile(std::filesystem::path path, bool remove)
+      : path(path), remove_(remove) {}
+
+  // Whether to remove the file on destruction or not.
+  bool remove_;
+};
+
+inline ErrorOr<CacheFile> CacheFile::create(const std::string &graphName,
+                                            const std::string &fileName,
+                                            bool remove) {
+  std::filesystem::path path = CacheFile::getPath(graphName, fileName);
+  FUSILLI_LOG_LABEL_ENDL("Createing Cache file");
+  FUSILLI_LOG_ENDL(path);
+
+  // Create directory $HOME/.cache/fusilli/<graphName>
+  std::filesystem::path cacheDir = path.parent_path();
+  std::error_code ec;
+  std::filesystem::create_directories(cacheDir, ec);
+  FUSILLI_RETURN_ERROR_IF(ec, ErrorCode::FileSystemFailure,
+                          "Failed to create cache directory: " +
+                              cacheDir.string() + " - " + ec.message());
+
+  // Create file $HOME/.cache/fusilli/<graphName>/<fileName>
+  std::ofstream file(path);
+  FUSILLI_RETURN_ERROR_IF(!file.is_open(), ErrorCode::FileSystemFailure,
+                          "Failed to create file: " + path.string());
+  file.close();
+
+  return ok(CacheFile(path, remove));
+}
+
+inline ErrorOr<CacheFile> CacheFile::open(const std::string &graphName,
+                                          const std::string &filename) {
+  std::filesystem::path path = CacheFile::getPath(graphName, filename);
+
+  // Check if the file exists.
+  FUSILLI_RETURN_ERROR_IF(!std::filesystem::exists(path),
+                          ErrorCode::FileSystemFailure,
+                          "File does not exist: " + path.string());
+
+  return ok(CacheFile(path, false));
+}
+
+inline std::filesystem::path CacheFile::getPath(const std::string &graphName,
+                                                const std::string &fileName) {
+  // Ensure graphName is safe to use as a directory name, we assume fileName
+  // is safe.
+  std::string sanitizedGraphName = graphName;
+  std::transform(sanitizedGraphName.begin(), sanitizedGraphName.end(),
+                 sanitizedGraphName.begin(),
+                 [](char c) { return c == ' ' ? '_' : c; });
+  std::erase_if(sanitizedGraphName,
+                [](unsigned char c) { return !(std::isalnum(c) || c == '_'); });
+
+  // Ensure graphName has a value.
+  if (sanitizedGraphName.empty()) {
+    sanitizedGraphName = "unnamed_graph";
+  }
```

Comment:
This is OK to have here, but let's avoid a can of worms by making a graphName mandatory on every Fusilli graph, so it is always provided (non-empty) to `CacheFile`.